api/monitor/monitorclient/client.go

```diff
@@ -0,0 +1,217 @@
+// Copyright © 2017 The Things Network
+// Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+
+package monitorclient
+
+import (
+	"context"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/TheThingsNetwork/go-utils/grpc/sendbuffer"
+	"github.com/TheThingsNetwork/go-utils/log"
+	"github.com/TheThingsNetwork/ttn/api/broker"
+	"github.com/TheThingsNetwork/ttn/api/gateway"
+	"github.com/TheThingsNetwork/ttn/api/handler"
+	"github.com/TheThingsNetwork/ttn/api/monitor"
+	"github.com/TheThingsNetwork/ttn/api/networkserver"
+	"github.com/TheThingsNetwork/ttn/api/pool"
+	"github.com/TheThingsNetwork/ttn/api/router"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// DefaultBufferSize sets the default buffer size per stream
+var DefaultBufferSize = 10
+
+// MonitorOption changes something in the MonitorClient
+type MonitorOption func(m *MonitorClient)
+
+// WithConn adds the given conns to the MonitorClient
+func WithConn(name string, conn *grpc.ClientConn) MonitorOption {
+	return func(m *MonitorClient) {
+		m.clients[name] = monitor.NewMonitorClient(conn)
+	}
+}
+
+// WithServer [DEPRECATED] connects to the given server and adds it to the MonitorClient
+// Instead of using WithServer, you should set up the gRPC connection externally and use
+// WithConn to add it to the MonitorClient.
+func WithServer(name, addr string, opts ...grpc.DialOption) MonitorOption {
+	if len(opts) == 0 {
+		if strings.HasSuffix(name, "-tls") {
+			netHost, _, _ := net.SplitHostPort(addr)
+			creds := credentials.NewTLS(pool.TLSConfig(netHost))
+			opts = append(opts, grpc.WithTransportCredentials(creds))
+		} else {
+			opts = append(opts, grpc.WithInsecure())
+		}
+	}
+	return func(m *MonitorClient) {
+		conn, err := grpc.Dial(addr, opts...)
+		if err != nil {
+			return
+		}
+		m.clients[name] = monitor.NewMonitorClient(conn)
+	}
+}
+
+// WithStatusTicker returns a MonitorOption that sets a status ticker
+func WithStatusTicker(interval time.Duration) MonitorOption {
+	return func(m *MonitorClient) {
+		m.statusTicker = time.NewTicker(interval)
+	}
+}
+
+// NewMonitorClient returns a new MonitorClient
+func NewMonitorClient(opts ...MonitorOption) *MonitorClient {
+	m := &MonitorClient{
+		log:        log.Get(),
+		bufferSize: DefaultBufferSize,
+		clients:    make(map[string]monitor.MonitorClient),
+	}
+	for _, opt := range opts {
+		opt(m)
+	}
+	return m
+}
+
+// MonitorClient is an overlay on top of the gRPC client
+type MonitorClient struct {
+	log          log.Interface
+	bufferSize   int
+	clients      map[string]monitor.MonitorClient
+	statusTicker *time.Ticker
+}
+
+// TickStatus calls f periodically if the StatusTicker has been set
+func (m *MonitorClient) TickStatus(f func()) {
```

Comment:
This func is for components that send their status on the stream every X seconds (for example: router, broker, networkserver, handler). I agree that it might be better to put this in the core